### DISCUSSION TITLE
hosting custom policies on server side

### DISCUSSION THREAD

**Discussion by bakuljajan** [aquasecurity/trivy#7172] [Ideas]
### Description

I'm only aware of the vuln database being supported to be hosted in trivy server. I would like to know if custom policies can also be hosted using trivy server. This should simplify distribution as we don't need to copy the rego files to each user.

### Target

Filesystem

### Scanner

Misconfiguration

**Comment by simar7** [comment]
Yes you can pass `--checks-bundle-repository` to a custom URL.

**Reply by bakuljajan** [reply]
But is that flag only available on `trivy config` right? Does that mean we must re-download the checks if it runs on VMs like on ci pipelines?

**Reply by simar7** [reply]
I'm not sure if I understand your question. Are you scanning for misconfigurations via `trivy  config`?  If not, how are you running trivy to scan for misconfigurations?

**Reply by nikpivkin** [reply]
@simar7 I think what the user means is that he is running Trivy in client-server mode and wants the checks to be stored on the server side

**Reply by bakuljajan** [reply]
Sorry @simar7 if my question is not clear enough, but it is exactly as @nikpivkin said, we want to use all features in trivy and we want to use client-server mode to mitigate having to redownload the vuln db every time the ci server is running to avoid rate limiting. 

As we explore more, we want to do some development on custom policies in misconfiguration check. But when I check on the documentation, the misconfiguration check via `trivy config` is not taking any server parameter and the `trivy server` command is not taking the `--config-check` and `--check-namespaces` parameters.

**Reply by knqyf263** [reply]
Misconfiguration scanning is performed on the client side. Otherwise, a client needs to send all the IaC files, which may include sensitive information, to the server.
Another benefit of the client/server mode is caching. In image scanning, the server stores the analysis result in the cache per layer, and the client can skip analyzing layers in subsequent scans. This is very useful for vulnerability scanning but not for misconfiguration scanning. In misconfiguration scanning, we need to cache the IaC file contents. In addition, filesystem scanning doesn't fit for cache.

For the above reasons, misconfiguration scanning (and secret scanning for the same reasons) doesn't support the client/server mode.

@simar7 @nikpivkin I think we should update [the document](https://aquasecurity.github.io/trivy/v0.53/docs/references/modes/client-server/).
- Removing AWS from the table
- Changing `Config` to unsupported in the table
- Adding a new table for supported scanners

**Reply by simar7** [reply]
Ah I see what you mean @bakuljajan -  @knqyf263 correct me if I'm wrong but in this case could you specify the experimental `--cache-backend` flag within trivy server to a custom path where you maintain it plus set up `--skip-check-update` within each trivy client to use the custom cache backend?

That way the the checks from the custom backend will be used and the checks bundle will not be downloaded by each of the trivy clients.

**Reply by simar7** [reply]
Sorry I just read your comment above @knqyf263 - it makes sense to me but I'm still curious if what I proposed could work or there are limitations to how a custom cache backend is used today?

**Reply by simar7** [reply]
Created https://github.com/aquasecurity/trivy/issues/7177 to update docs.

**Reply by knqyf263** [reply]
There are two types of artifacts in the cache directory today.

1. Scan cache (analysis cache)
    - Store analysis results to speed up subsequent scans
        - Vulnerability scanner: A list of installed packages (name, version, license, etc.)
        - Misconfiguration scanner: Misconfiguration results
        - Secret scanner: Masked secret results
        - License scanner: Same as the vulnerability scanner
2. DB, etc.
    - Vulnerability DB
    - Java DB
    - Checks Bundle

`--cache-backend` is actually `--scan-cache-backend` (we might want to rename it for clarity). The flag doesn't affect the DB and bundle.

**Comment by nikpivkin** [comment]
We will change the documentation to clarify some details when scanning for misconfigurations in client/server mode.

Track https://github.com/aquasecurity/trivy/issues/7177
